### PR TITLE
DBZ-5283 Introduce a new SMT ExtractChangedRecordState to capture the names of changed/unchanged field within headers

### PR DESCRIPTION
**Pull request by harveyyue**
https://issues.redhat.com/browse/DBZ-5283

### CODE REVIEW

**Comment by jpechane** [comment]
@harveyyue Thanks for the PR! I'd have few recommendations for improvement
1) I'd prefer this as a separate SMT that will operate on the envelope. That way the people who don't want to use flattening can still benefit from getting only changed fields in before/after part.
2) There shoud be two headers - one enumerating changed and the other unchanged fields
3) The names of both headers will be configurable with `null` as default. If set to something the info for the given header will be provided.

WDYT?

**Comment by harveyyue** [comment]
@jpechane Agree with your recommendations.

**Comment by jpechane** [comment]
@harveyyue Thanks for the PR, it is closer to what I was looking for. Still I'd like to ask for few more improvements

* Please take a look at `io.debezium.transforms.SmtManager`, it might be useful to be used for this SMT as well
* The headers were inteded to be optional, so if a config option is not set then the apporpriate header is not sent. The current implementation sends headers always with just a default name
* I'd like to have one additional functionality - not only enumerating headers but if enabled be a config option only fields that has changed will be kept in both `before` and `after` fields so effectively diff

**Comment by harveyyue** [comment]
@jpechane 
point 2: Do you mean if not set changed/unchanged header name we should not send to headers?
point 3: We should add a switch config to identify only send update event changed/unchanged fields to headers, otherwise handle all events(c/u/d), right?

**Comment by jpechane** [comment]
@harveyyue For point 2, yes exactly. For point 3 - here I am talking not about headers but the message payload. Only changed fields should be kept inbefore and after payload.

**Comment by harveyyue** [comment]
@jpechane Got it, I will complete these point at this weekend.

**Comment by jpechane** [comment]
@Naros @harveyyue I still believe this PR is incomplete. The original issue was about providing update (patch) transformation that will altert the payload to keep only updated fields and add metadata. Right now it does the metadata part but not the payload part. WDYT?

**Comment by Naros** [comment]
I'll look at the payload part today, I must have overlooked that bit @jpechane.

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+** The operation that made the change
+** Source information such as the names of the database and table where the change was made
+** Time stamp for when the change was made
+** Optional transaction information
+
+* Row data before the change
+* Row data after the change
+
+For example, part of the structure of an `UPDATE` change event looks like this:
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:43]:
```suggestion
The following example shows part of the structure of a typical {prodname} `UPDATE` change event:
```

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+However, other connectors or other parts of the Kafka ecosystem may expect knowledge of what fields changed or are unchanged, and this event changes SMT allows exposing that information via event headers.
+
+[[event-changes-behavior]]
+== Behavior
+
+The event changes SMT extracts the `before` and `after` fields from a {prodname} `UPDATE` change event in a Kafka record.
+The SMT will examine each of the fields within these two  event state structures and writes the names of fields that changed or are unchanged based on the configuration supplied to the SMT.
+If the event represents an `INSERT` or `DELETE`, this single message transformation has no effect.
+
+You can configure the event changes SMT for a {prodname} connector or for a sink connector that consumes messages emitted by a {prodname} connector.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:79]:
```suggestion
You can configure the event changes SMT for a {prodname} connector, or for a sink connector that consumes messages emitted by a {prodname} connector.
```

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+The advantage of configuring the event changes for a sink connector is that the records stored in Apache Kafka contain whole {prodname} change events.
+The decision to apply the SMT to a source or sink connector depends on your particular use case.
+
+You can configure the transformation to do any of the following:
+
+* Record the fields that changed in the `UPDATE` in a user-configured header.
+* Record the fields that did not change in the `UPDATE` in a user-configured header.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:86]:
```suggestion
* Identify the fields that are changed by an `UPDATE` event by listing them in the user-configured xref:extract-changes-header-changed-name[`header.changed.name`] header.
* Identify the fields that are not changed by an `UPDATE` event by listing them in the user-configured xref:extract-changes-header-unchanged-name[`header.unchanged.name`] header.
```

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+[[event-changes-behavior]]
+== Behavior
+
+The event changes SMT extracts the `before` and `after` fields from a {prodname} `UPDATE` change event in a Kafka record.
+The SMT will examine each of the fields within these two  event state structures and writes the names of fields that changed or are unchanged based on the configuration supplied to the SMT.
+If the event represents an `INSERT` or `DELETE`, this single message transformation has no effect.
+
+You can configure the event changes SMT for a {prodname} connector or for a sink connector that consumes messages emitted by a {prodname} connector.
+The advantage of configuring the event changes for a sink connector is that the records stored in Apache Kafka contain whole {prodname} change events.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:80]:
```suggestion
Configure the event changes SMT for a sink connector if you want Apache Kafka to retain the entire original {prodname} change events .
```

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+
+{prodname} generates data change events that have a complex structure.
+Each event consists of three parts:
+
+* Metadata, which includes but is not limited to:
+
+** The operation that made the change
+** Source information such as the names of the database and table where the change was made
+** Time stamp for when the change was made
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:37]:
```suggestion
** The type of operation that changed the data.
** Source information, such as the names of the database and the table in which the change occurred.
** Timestamp that identifies when the change was made.
```

Comment (roldanbob) [reply documentation/modules/ROOT/pages/transformations/event-changes.adoc:37]: I know that this same list appears elsewhere (e.g., the event-flattening SMT docs). I'm in the process of making similar changes there as well.

**Review by roldanbob — CHANGES_REQUESTED** [review]

**Comment by Naros** [comment]
As outlined on Zulip, I added all but the `null.unchanged.values` suggestion from the Jira because I have an open question regarding the event's schema evolution in doing so; otherwise I believe this is now good.

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+Configure the {prodname} event changes SMT in a Kafka Connect source or sink connector by adding the SMT configuration details to your connector's configuration.
+To obtain the default behavior which doesn't add any headers, in a `.properties` file, you would specify something like the following:
+
+[source]
+----
+transforms=changes,...
+transforms.changes.type=io.debezium.transforms.ExtractChangedRecordState
+----
+
+As with any Kafka Connect connector configuration, you can set `transforms=` to multiple, comma-seperated, SMT aliases in the order in which you want Kafka Connect to apply the SMTs.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:102]:
```suggestion
As with any Kafka Connect connector configuration, you can set `transforms=` to multiple, comma-separated, SMT aliases in the order in which you want Kafka Connect to apply the SMTs.
```

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc [hunk 7]
```diff
@@ -304,3 +304,32 @@ When the SMT adds metadata fields to the simplified record's header, it prefixes
...
+
+
+.Descriptions of event flattening SMT configuration options for integration with event changes SMT
+[cols="30%a,25%a,45%a",subs="+attributes",options="header"]
+|===
+|Option
+|Default
+|Description
+
+|[[extract-new-record-state-header-changed-name]]xref:{link-event-flattening}#extract-new-record-state-header-unchanged-name[`header.unchanged.name`]
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-flattening.adoc:318]:
```suggestion
|[[extract-new-record-state-header-changed-name]]xref:extract-new-record-state-header-unchanged-name[`header.unchanged.name`]
```
There's no need to include the category attribute for xrefs that link to content in the same file. Same comment applies to the table entries that follow.

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc [hunk 8]
```diff
@@ -304,3 +304,32 @@ When the SMT adds metadata fields to the simplified record's header, it prefixes
...
+|
+|The Kafka message header to source unchanged field names from.
+
+[id="extract-new-record-state-drop-unchanged-fields"]
+|xref:{link-event-flattening}#extract-new-record-state-drop-unchanged-fields[`drop.unchanged.fields`]
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-flattening.adoc:323]:
```suggestion
|xref:extract-new-record-state-drop-unchanged-fields[`drop.unchanged.fields`]
```

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc [hunk 9]
```diff
@@ -304,3 +304,32 @@ When the SMT adds metadata fields to the simplified record's header, it prefixes
...
+|`false`
+|{prodname} generates a flattened event structure based on the `after` portion of the event's payload that includes all fields by default.
+For an `UPDATE` operation, this behavior can be adjusted to only include the fields that were changed by setting this option to `true` to drop or omit fields that were not changed in the update.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-flattening.adoc:326]:
```suggestion
For an `UPDATE` operation, set this option to `true` if you want the SMT to only include fields that were changed during an operation. Fields that were not changed in the update are dropped. 
```

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc [hunk 10]
```diff
@@ -304,3 +304,32 @@ When the SMT adds metadata fields to the simplified record's header, it prefixes
...
+
+[id="extract-new-record-state-keep-unchanged-primary-keys"]
+|xref:{link-event-flattening}#extract-new-record-state-keep-unchanged-primary-keys[`keep.unchanged.primary.keys`]
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-flattening.adoc:329]:
```suggestion
|xref:extract-new-record-state-keep-unchanged-primary-keys[`keep.unchanged.primary.keys`]
```

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc [hunk 11]
```diff
@@ -304,3 +304,32 @@ When the SMT adds metadata fields to the simplified record's header, it prefixes
...
+|`true`
+|{prodname} generates a flattened event structure based on the `after` portion of the event's payload that includes all fields by default.
+When enabling link:#extract-new-record-state-drop-unchanged-fields[`drop.unchanged.fields`], it is possible to omit primary key values in the flattened structure by setting this option to `false`.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-flattening.adoc:332]:
```suggestion
If xref:extract-new-record-state-drop-unchanged-fields[`drop.unchanged.fields`] is set to `true`, set `keep.unchanged.primary.keys` to `false` to omit primary key values from the flattened structure.
```

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc [hunk 12]
```diff
@@ -304,3 +304,32 @@ When the SMT adds metadata fields to the simplified record's header, it prefixes
...
+By default, primary key fields are always retained even if unchanged fields are to be omitted.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-flattening.adoc:333]:
```suggestion
The default setting (`true`) retains primary key fields even if the event changes SMT (`ExtractChangedRecordState`) is configured to omit unchanged fields.
```

**Review by roldanbob — CHANGES_REQUESTED** [review]

**Comment by jpechane** [comment]
@Naros We are getting there but I think there is a misunderstanding. My intent was not to couple those two SMTs.
The idea was

1. Extend `ExtractNewRecordState` with a functionality to remove arbitrary fields for the message. The list is passed as a header. It is not about being unchange/changed etc. just a list of columns in a configured header. So the code should reflect this nature and not be bound to unchanged/changed concept.
2. A new SMT (which is more or less done here) will just prepare the list of unchanged fields. So there is only loose relation between those two SMTs.

Schema evolution
I was thinking about that and I beleive that `ExtractNewRecordState` should operate in two modes

- schema incompatible - the fields will just get removed from both schema and payload
- schema compatible - here the optional fields will not be provided. I think we should keep mandatory fields in place as it is in my opinion the pth of least resistance.

WDYT?

**Comment by Naros** [comment]
@roldanbob can you give the latest commit a review?  I adjusted the properties based on discussions with Jiri so your edits are now a tad out of date.

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc [hunk 13]
```diff
@@ -66,9 +66,10 @@ More details about change event structure are provided in
 xref:{link-connectors}[the documentation for each connector].
 endif::community[]
 
-This complex format provides the most information about changes happening in the system.
-However, downstream consumers of the message (other connectors, or other parts of the Kafka ecosystem) might expect the message to explicitly identify the fields that a database operation changes or leaves unchanged.
-This event changes SMT adds information about these field changes to headers in the change event message.
+The complex format of the message in the preceding example provides detailed information about changes that occur in the source database.
+However, the format might not be suitable for some downstream consumers.
+Sink connectors, or other parts of the Kafka ecosystem might expect the message to explicitly identify the fields that a database operation changes or leaves unchanged.
+The `ExtractChangedRecordState` SMT adds headers to the change event message to identify the fields that are modified by a database operation or left unchanged.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:72]:
```suggestion
The `ExtractChangedRecordState` SMT adds headers to the change event message to identify the fields that are modified by a database operation, and the fields that remain unchanged.
```

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc [hunk 14]
```diff
@@ -81,13 +82,11 @@ You can configure the event changes SMT for a {prodname} connector, or for a sin
 Configure the event changes SMT for a sink connector if you want Apache Kafka to retain the entire original {prodname} change events.
 The decision to apply the SMT to a source or sink connector depends on your particular use case.
 
-You can configure the transformation to do any of the following:
+Depending on your use case, you can configure the transformation to modify the original message by performing one of both of the following tasks:
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:85]:
```suggestion
Depending on your use case, you can configure the transformation to modify the original message by performing one or both of the following tasks:
```

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc [hunk 15]
```diff
@@ -303,33 +303,22 @@ Optionally, you can override the field name via `<field name>:<new field name>`,
...
-|Option
-|Default
-|Description
 
-|[[extract-new-record-state-header-changed-name]]xref:{link-event-flattening}#extract-new-record-state-header-unchanged-name[`header.unchanged.name`]
+[id="extract-new-record-state-drop-fields-header-name"]
+|xref:{link-event-flattening}#extract-new-record-state-drop-fields-header-name[`drop.fields.header.name`]
 |
-|The Kafka message header to source unchanged field names from.
+|The Kafka message header name to source fields that should be dropped.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-flattening.adoc:310]:
```suggestion
|The Kafka message header name to use for listing field names in the source message that you want to drop from the output message.
```
Does this specify the name of  the header that you want to use to list the fields in the source message that you want to drop in the output message? Or does it represent the name(s) of the headers in the source that you want to drop from the output message?

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc [hunk 16]
```diff
@@ -303,33 +303,22 @@ Optionally, you can override the field name via `<field name>:<new field name>`,
...
 
-[id="extract-new-record-state-drop-unchanged-fields"]
-|xref:{link-event-flattening}#extract-new-record-state-drop-unchanged-fields[`drop.unchanged.fields`]
+[id="extract-new-record-state-drop-fields-from-key"]
+|xref:{link-event-flattening}#extract-new-record-state-drop-fields-from-key[`drop.fields.from.key`]
 |`false`
-|{prodname} generates a flattened event structure based on the `after` portion of the event's payload that includes all fields by default.
-For an `UPDATE` operation, this behavior can be adjusted to only include the fields that were changed by setting this option to `true` to drop or omit fields that were not changed in the update.
+|Controls whether fields are also be dropped fom the event's key.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-flattening.adoc:315]:
```suggestion
|Specifies whether you want the SMT to remove fields that are listed in `drop.fields.header.name` from the event's key.
```
I'm assuming that the fields to remove from the key are those that you specify in `drop.fields.header.name`?

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc [hunk 17]
```diff
@@ -303,33 +303,22 @@ Optionally, you can override the field name via `<field name>:<new field name>`,
...
 
-[id="extract-new-record-state-keep-unchanged-primary-keys"]
-|xref:{link-event-flattening}#extract-new-record-state-keep-unchanged-primary-keys[`keep.unchanged.primary.keys`]
+[id="extract-new-record-state-drop-fields-keep-schema-compatible"]
+|xref:{link-event-flattening}#extract-new-record-state-drop-fields-keep-schema-compatible[`drop.fields.keep.schema.compatible`]
 |`true`
-|{prodname} generates a flattened event structure based on the `after` portion of the event's payload that includes all fields by default.
-When enabling link:#extract-new-record-state-drop-unchanged-fields[`drop.unchanged.fields`], it is possible to omit primary key values in the flattened structure by setting this option to `false`.
-By default, primary key fields are always retained even if unchanged fields are to be omitted.
+|Controls whether non-optional fields will be removed if they're included in the xref:{link-event-flattening}#extract-new-record-state-drop-fields-header-name[`drop.fields.header.name`] configuration property. +
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-flattening.adoc:320]:
```suggestion
|Specifies whether you want the SMT to remove non-optional fields that are included in the xref:extract-new-record-state-drop-fields-header-name[`drop.fields.header.name`] configuration property. +
```

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc [hunk 18]
```diff
@@ -303,33 +303,22 @@ Optionally, you can override the field name via `<field name>:<new field name>`,
...
-.Descriptions of event flattening SMT configuration options for integration with event changes SMT
-[cols="30%a,25%a,45%a",subs="+attributes",options="header"]
-|===
-|Option
-|Default
-|Description
 
-|[[extract-new-record-state-header-changed-name]]xref:{link-event-flattening}#extract-new-record-state-header-unchanged-name[`header.unchanged.name`]
+[id="extract-new-record-state-drop-fields-header-name"]
+|xref:{link-event-flattening}#extract-new-record-state-drop-fields-header-name[`drop.fields.header.name`]
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-flattening.adoc:308]:
```suggestion
|xref:extract-new-record-state-drop-fields-header-name[`drop.fields.header.name`]
```

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc [hunk 19]
```diff
@@ -303,33 +303,22 @@ Optionally, you can override the field name via `<field name>:<new field name>`,
...
 |
-|The Kafka message header to source unchanged field names from.
+|The Kafka message header name to source fields that should be dropped.
 
-[id="extract-new-record-state-drop-unchanged-fields"]
-|xref:{link-event-flattening}#extract-new-record-state-drop-unchanged-fields[`drop.unchanged.fields`]
+[id="extract-new-record-state-drop-fields-from-key"]
+|xref:{link-event-flattening}#extract-new-record-state-drop-fields-from-key[`drop.fields.from.key`]
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-flattening.adoc:313]:
```suggestion
|xref:extract-new-record-state-drop-fields-from-key[`drop.fields.from.key`]
```

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc [hunk 20]
```diff
@@ -303,33 +303,22 @@ Optionally, you can override the field name via `<field name>:<new field name>`,
...
 |`false`
-|{prodname} generates a flattened event structure based on the `after` portion of the event's payload that includes all fields by default.
-For an `UPDATE` operation, this behavior can be adjusted to only include the fields that were changed by setting this option to `true` to drop or omit fields that were not changed in the update.
+|Controls whether fields are also be dropped fom the event's key.
 
-[id="extract-new-record-state-keep-unchanged-primary-keys"]
-|xref:{link-event-flattening}#extract-new-record-state-keep-unchanged-primary-keys[`keep.unchanged.primary.keys`]
+[id="extract-new-record-state-drop-fields-keep-schema-compatible"]
+|xref:{link-event-flattening}#extract-new-record-state-drop-fields-keep-schema-compatible[`drop.fields.keep.schema.compatible`]
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-flattening.adoc:318]:
```suggestion
|xref:extract-new-record-state-drop-fields-keep-schema-compatible[`drop.fields.keep.schema.compatible`]
```

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc [hunk 21]
```diff
@@ -303,33 +303,22 @@ Optionally, you can override the field name via `<field name>:<new field name>`,
...
 |`true`
-|{prodname} generates a flattened event structure based on the `after` portion of the event's payload that includes all fields by default.
-When enabling link:#extract-new-record-state-drop-unchanged-fields[`drop.unchanged.fields`], it is possible to omit primary key values in the flattened structure by setting this option to `false`.
-By default, primary key fields are always retained even if unchanged fields are to be omitted.
+|Controls whether non-optional fields will be removed if they're included in the xref:{link-event-flattening}#extract-new-record-state-drop-fields-header-name[`drop.fields.header.name`] configuration property. +
+ +
+The default behavior is to only remove fields if they're marked optional.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-flattening.adoc:322]:
```suggestion
By default, the SMT only removes fields that are marked `optional`.
```

**Review by roldanbob — CHANGES_REQUESTED** [review]

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc (hunk 15 above)
Comment (Naros) [reply documentation/modules/ROOT/pages/transformations/event-flattening.adoc:310]: The fields, so I believe your suggestion is correct.

File: documentation/modules/ROOT/pages/transformations/event-flattening.adoc (hunk 16 above)
Comment (Naros) [reply documentation/modules/ROOT/pages/transformations/event-flattening.adoc:315]: Correct.

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+
+Depending on your use case, you can configure the transformation to modify the original message by performing one or both of the following tasks:
+
+* Identify the fields that are changed by an `UPDATE` event by listing them in the user-configured xref:extract-changes-header-changed-name[`header.changed.name`] header.
+* Identify the fields that are not changed by an `UPDATE` event by listing them in the user-configured xref:extract-changes-header-unchanged-name[`header.unchanged.name`] header.
+
+== Configuration
+
+Configure the {prodname} event changes SMT in a Kafka Connect source or sink connector by adding the SMT configuration details to your connector's configuration.
+To obtain the default behavior which doesn't add any headers, in a `.properties` file, you would specify something like the following:
```
Comment (jpechane) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:93]: I don't think we should talk about `.properties` file. It is connector configuration.

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+
+[source]
+----
+transforms=changes,...
+transforms.changes.type=io.debezium.transforms.ExtractChangedRecordState
+----
+
+As with any Kafka Connect connector configuration, you can set `transforms=` to multiple, comma-separated, SMT aliases in the order in which you want Kafka Connect to apply the SMTs.
+
+The following `.properties` example sets several event changes SMT options:
```
Comment (jpechane) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:103]: See above

**Review by jpechane — APPROVED** [review]
@Naros LGTM, thanks! I left two minor comments related to docs. When you and @roldanbob agree on the docs fully feel free to merge the PR yourself.

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc (hunk 22 above)
Comment (roldanbob) [reply documentation/modules/ROOT/pages/transformations/event-changes.adoc:93]: This is copied in from something that I introduced in some of the other transformation docs a couple of years ago. It appears in most of the other docs. We can remove it here and on L103 @Naros, and I'll open a Jira to correct instances that appear elsewhere.

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc (hunk 23 above)
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:103]:
```suggestion
The connector configuration in the following example sets several options for the event changes SMT:
```

**Review by roldanbob — CHANGES_REQUESTED** [review]

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc (hunk 22 above)
[the remark this answers is not in the document]
Comment (roldanbob) [reply documentation/modules/ROOT/pages/transformations/event-changes.adoc:93]: Removes reference on L93 to `.properties` file per earlier comment.

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc (hunk 23 above)
Comment (roldanbob) [reply documentation/modules/ROOT/pages/transformations/event-changes.adoc:103]: Removes reference on L103 to `.properties` file per earlier comment.

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc (hunk 22 above)
Comment (Naros) [reply documentation/modules/ROOT/pages/transformations/event-changes.adoc:93]: I believe these are just copied from the `ExtractNewRecordState` SMT documentation and amended to be relevant to this SMT, but the formatting is relatively the same.  I see Bob has provided some edits below, I'll adjust accordingly.

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+.Descriptions of event changes SMT configuration options
+[cols="30%a,25%a,45%a",subs="+attributes",options="header"]
+|===
+|Option
+|Default
+|Description
+
+|[[extract-changes-header-changed-name]]<<extract-changes-header-changed-name, `+header.changed.name+`>>
+|
+|Specifies the name of the header that will include a comma-separated list of fields that were changed.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:143]:
```suggestion
|The Kafka message header name to use for storing a comma-separated list of the fields that are changed by a database operation.
```

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+
+|[[extract-changes-header-unchanged-name]]<<extract-changes-header-unchanged-name, `+header.unchanged.name+`>>
+|
+|Specifies the name of the header that will include a comma-separated list of fields that were not changed.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:147]:
```suggestion
|The Kafka message header name to use for storing a comma-separated list of the fields that remain unchanged after a database operation.
```

**Review by roldanbob — CHANGES_REQUESTED** [review]

**Comment by Naros** [comment]
Once we're good on the docs @roldanbob, I would like to squash all these multiple commits, so could we confirm that the docs are in a good spot so I can do that?

**Comment by roldanbob** [comment]
> Once we're good on the docs @roldanbob, I would like to squash all these multiple commits, so could we confirm that the docs are in a good spot so I can do that?

Should be good ... Taking a look now to confirm and will let you know within the hour.

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+The complex format of the message in the preceding example provides detailed information about changes that occur in the source database.
+However, the format might not be suitable for some downstream consumers.
+Sink connectors, or other parts of the Kafka ecosystem might expect the message to explicitly identify the fields that a database operation changes or leaves unchanged.
+The `ExtractChangedRecordState` SMT adds headers to the change event message to identify the fields that are modified by a database operation, and the fields that remain unchanged.
+
+[[event-changes-behavior]]
+== Behavior
+
+The event changes SMT extracts the `before` and `after` fields from a {prodname} `UPDATE` change event in a Kafka record.
+The SMT examines the fields in the `before` and `after` event state structures, and then writes the names of fields that changed or are unchanged based on the values of  settings that you apply to the SMT configuration.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/transformations/event-changes.adoc:78]:
```suggestion
The transformation examines the `before` and `after` event state structures to identify the fields that are altered by an operation, and those that remain unchanged.
Depending on the connector configuration, the transformation then produces a modified event message that adds message headers to list the changed fields, the unchanged fields, or both.
```

**Review by roldanbob — CHANGES_REQUESTED** [review]

File: documentation/modules/ROOT/pages/transformations/event-changes.adoc (hunk 26 above)
Comment (roldanbob) [reply documentation/modules/ROOT/pages/transformations/event-changes.adoc:78]: @Naros A few small wording changes to further unpack the meaning. If this looks ok to you, I'm good with releasing this to the wild.

**Review by Naros — APPROVED** [review]

**Comment by jpechane** [comment]
Applied, thanks to all!

**Pull request merged**